Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift

```diff
@@ -187,4 +202,185 @@ class HTTP2ClientTests: XCTestCase {
+        var response1: ResponseAccumulator.Response?
+        XCTAssertNoThrow(response1 = try task1.wait())
+
+        XCTAssertEqual(.ok, response1?.status)
+        XCTAssertEqual(response1?.version, .http2)
+        let serverPort = bin.port
+        XCTAssertNoThrow(try bin.shutdown())
+        // client is now in HTTP/2 state and the HTTPBin is closed
+        // start a new server on the old port which closes all connections immediately
+        let serverGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        defer { XCTAssertNoThrow(try serverGroup.syncShutdownGracefully()) }
+        var maybeServer: Channel?
+        XCTAssertNoThrow(maybeServer = try ServerBootstrap(group: serverGroup)
+            .serverChannelOption(ChannelOptions.socketOption(.so_reuseaddr), value: 1)
+            .childChannelInitializer { channel in
+                channel.close()
+            }
+            .childChannelOption(ChannelOptions.socketOption(.so_reuseaddr), value: 1)
+            .bind(host: "0.0.0.0", port: serverPort)
+            .wait())
+        guard let server = maybeServer else { return }
+        defer { XCTAssertNoThrow(try server.close().wait()) }
+
+        var maybeRequest2: HTTPClient.Request?
+        XCTAssertNoThrow(maybeRequest2 = try HTTPClient.Request(url: "https://localhost:\(serverPort)/"))
+        guard let request2 = maybeRequest2 else { return }
+
+        let task2 = client.execute(request: request2, delegate: ResponseAccumulator(request: request2), eventLoop: .delegateAndChannel(on: el2))
+        XCTAssertThrowsError(try task2.wait()) { error in
+            XCTAssertNil(
+                error as? HTTPClientError,
+                "error should be some platform specific error that the connection is closed/reset by the other side"
+            )
+        }
+    }
+}
+
+private final class HeadReceivedCallback: HTTPClientResponseDelegate {
+    typealias Response = Void
+    private let didReceiveHeadCallback: (HTTPResponseHead) -> Void
+    init(didReceiveHead: @escaping (HTTPResponseHead) -> Void) {
+        self.didReceiveHeadCallback = didReceiveHead
+    }
+
+    func didReceiveHead(task: HTTPClient.Task<Void>, _ head: HTTPResponseHead) -> EventLoopFuture<Void> {
+        self.didReceiveHeadCallback(head)
+        return task.eventLoop.makeSucceededVoidFuture()
+    }
+
+    func didFinishRequest(task: HTTPClient.Task<Void>) throws {}
+}
+
+/// sends some headers and waits indefinitely afterwards
+private final class SendHeaderAndWaitChannelHandler: ChannelInboundHandler {
+    typealias InboundIn = HTTPServerRequestPart
+    typealias OutboundOut = HTTPServerResponsePart
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        context.writeAndFlush(wrapOutboundOut(.head(HTTPResponseHead(
+            version: HTTPVersion(major: 1, minor: 1),
+            status: .ok
+        ))
+        ), promise: nil)
```

Comment:
This will send at least two response headers for each request... you'll need to unwrap the incoming requests for `.head` and `.end`

Reply:
could we also use `channelActive` instead?

Reply:
nope... you need to wait for the request.

Reply:
okay thanks, we now only send it after we received `.end`.